reference/test/matrix/batch_diagonal_kernels.cpp

```diff
@@ -0,0 +1,502 @@
+    GKO_ASSERT_MTX_NEAR(umtx[0].get(), ures[0].get(), residual);
+    GKO_ASSERT_MTX_NEAR(umtx[1].get(), ures[1].get(), residual);
+}
+
+
+TYPED_TEST(BatchDiagonal, MovesToPrecision)
+{
+    using BatchDiagonal = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    using OtherT = typename gko::next_precision<T>;
+    using OtherBatchDiagonal = typename gko::matrix::BatchDiagonal<OtherT>;
+    auto tmp = OtherBatchDiagonal::create(this->exec);
+    auto res = BatchDiagonal::create(this->exec);
+    // If OtherT is more precise: 0, otherwise r
+    auto residual = r<OtherT>::value < r<T>::value
+                        ? gko::remove_complex<T>{0}
+                        : gko::remove_complex<T>{r<OtherT>::value};
+
+    this->mtx_1->move_to(tmp.get());
+    tmp->move_to(res.get());
+
+    auto ures = res->unbatch();
+    auto umtx = this->mtx_1->unbatch();
+    GKO_ASSERT_MTX_NEAR(umtx[0].get(), ures[0].get(), residual);
+    GKO_ASSERT_MTX_NEAR(umtx[1].get(), ures[1].get(), residual);
+}
+
+
+TYPED_TEST(BatchDiagonal, ConvertsEmptyToPrecision)
+{
+    using BatchDiagonal = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    using OtherT = typename gko::next_precision<T>;
+    using OtherBatchDiagonal = typename gko::matrix::BatchDiagonal<OtherT>;
+    auto empty = OtherBatchDiagonal::create(this->exec);
+    auto res = BatchDiagonal::create(this->exec);
+
+    empty->convert_to(res.get());
+
+    ASSERT_FALSE(res->get_num_batch_entries());
+}
+
+
+TYPED_TEST(BatchDiagonal, MovesEmptyToPrecision)
+{
+    using BatchDiagonal = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    using OtherT = typename gko::next_precision<T>;
+    using OtherBatchDiagonal = typename gko::matrix::BatchDiagonal<OtherT>;
+    auto empty = OtherBatchDiagonal::create(this->exec);
+    auto res = BatchDiagonal::create(this->exec);
+
+    empty->move_to(res.get());
+
+    ASSERT_FALSE(res->get_num_batch_entries());
+}
+
+
+TYPED_TEST(BatchDiagonal, SquareMatrixIsTransposable)
+{
+    using Mtx = typename TestFixture::Mtx;
+
+    auto trans = this->mtx_1->transpose();
+    auto trans_as_batch_diag = static_cast<Mtx*>(trans.get());
+
+    GKO_ASSERT_BATCH_MTX_NEAR(trans_as_batch_diag, this->mtx_1, 0.0);
+}
+
+
+TYPED_TEST(BatchDiagonal, NonSquareMatrixIsTransposable)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+
+    auto trans = this->mtx_5->transpose();
+    auto trans_as_batch_diag = static_cast<Mtx*>(trans.get());
+    auto check =
+        Mtx::create(this->exec, gko::transpose(this->mtx_5->get_size()));
+    this->exec->copy(this->mtx_5->get_num_stored_elements(),
+                     this->mtx_5->get_const_values(), check->get_values());
+
+    GKO_ASSERT_BATCH_MTX_NEAR(trans_as_batch_diag, check, 0.0);
+}
+
+
+using namespace std::complex_literals;
+
+
+template <typename T>
+class ComplexBatchDiagonal : public ::testing::Test {
```

Comment:
Nice that you test with non-trivial complex values, but I think without loss of generality, we could also directly test it in the above framework without a specific complex test suite ?

Reply:
I would then be creating the complex-valued matrices within the individual test functions, and not using the class member matrices. Or I would add the complex-valued matrices to the larger test suite class. But I do want to test at least conjugate transpose, and preferably transpose, with actual complex values.